include/sparrow/any_data.hpp

```diff
@@ -0,0 +1,349 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <any>
+#include <ranges>
+#include <tuple>
+#include <type_traits>
+#include <typeindex>
+#include <vector>
+
+#include "sparrow/any_data_utils.hpp"
+
+namespace sparrow
+{
+    /// A class that can own or not any object and expose it as a raw pointer.
+    template <typename T>
+    class any_data
+    {
+    public:
+
+        explicit any_data(T* data);
+
+        template <mpl::unique_ptr_or_derived U>
+            requires std::is_same_v<typename U::element_type, T>
+        explicit any_data(U data);
+
+        template <mpl::shared_ptr_or_derived U>
+            requires std::is_same_v<typename U::element_type, T>
+        explicit any_data(U data);
+
+
+        explicit any_data(T data)
+            requires(!mpl::smart_ptr_and_derived<T>);
+
+        T* get();
+        const T* get() const;
+
+        // Returns the data as a reference to the type U.
+        // The type U must be the same as the type of the data. Otherwise std::bad_any_cast exception is
+        // thrown.
```

Comment:
This comment is misleading, it suggests that `get_data<const int>` will return a `const int&` object, while the `get_data` return by value. I see two options here:
- either update the comment to reflect the code
- fix the code by returning `U&` and `const U&`. We should choose this option if returning by value is explicitely forbidden.

Reply:
I modified the comment